cpu/sam0_common: ADC: add support for samd5x/same5x

janosbrodbeck opened the pull request:
<!--
The RIOT community cares a lot about code quality.
Therefore, before describing what your contribution is about, we would like
you to make sure that your modifications are compliant with the RIOT
coding conventions, see https://github.com/RIOT-OS/RIOT/wiki/Coding-conventions.
-->

### Contribution description
This PR adds ADC support for the samd5x/same5x CPU family. Those CPUs have two ADC devices: ADC0 and ADC1. To support both devices this PR adds an ADC_DEV alias for the respective device (ADC/ADC0/ADC1). By default the driver uses ADC0 as device but it can be changed to ADC1 by a define (`#define ADC_DEV ADC1`). 

It is now also possible to change the ADC clock source via `#define ADC_GCLK_SRC {}`. The default is `SAM0_GCLK_MAIN`.

Also configured the same54-xpro board to make use of these changes.
### Testing procedure
**tests/periph_adc**
Flash the test on your board. It should work with all devices supported so far and additionally with same5x/samd5x MCUs (e.g. same54-xpro). 
<!--
Details steps to test your contribution:
- which test/example to compile for which board and is there a 'test' command
- how to know that it was not working/available in master
- the expected success test output
-->


### Issues/PRs references

<!--
Examples: Fixes #1234. See also #5678. Depends on PR #9876.

Please use keywords (e.g., fixes, resolve) with the links to the issues you
resolved, this way they will be automatically closed when your pull request
is merged. See https://help.github.com/articles/closing-issues-using-keywords/.
-->

benpicco commented on cpu/sam0_common/periph/adc.c:
```diff
@@ -120,33 +134,73 @@ static int _adc_configure(adc_res_t res)
     /* Power on */
 #ifdef CPU_SAML1X
     MCLK->APBCMASK.reg |= MCLK_APBCMASK_ADC;
+#endif
+#ifdef MCLK_APBDMASK_ADC0
+    if (ADC_DEV == ADC0) {
+        MCLK->APBDMASK.reg |= MCLK_APBDMASK_ADC0;
+    } else {
+        MCLK->APBDMASK.reg |= MCLK_APBDMASK_ADC1;
+    }
 #else
     MCLK->APBDMASK.reg |= MCLK_APBDMASK_ADC;
 #endif
+
+#ifdef ADC0_GCLK_ID
     /* GCLK Setup */
-    GCLK->PCHCTRL[ADC_GCLK_ID].reg = GCLK_PCHCTRL_CHEN | GCLK_PCHCTRL_GEN_GCLK0;
-    /* Set Voltage Reference */
-    ADC->REFCTRL.reg = ADC_REF_DEFAULT;
+    if (ADC_DEV == ADC0) {
+        GCLK->PCHCTRL[ADC0_GCLK_ID].reg = GCLK_PCHCTRL_CHEN |
+            GCLK_PCHCTRL_GEN(ADC_GCLK_SRC);
+    }
+    else {
+        GCLK->PCHCTRL[ADC1_GCLK_ID].reg = GCLK_PCHCTRL_CHEN |
+            GCLK_PCHCTRL_GEN(ADC_GCLK_SRC);
+    }
+    /* Configure CTRLA & CTRLB Register */
+    ADC_DEV->CTRLA.reg = ADC_PRESCALER;
+    ADC_DEV->CTRLB.reg |= res;
+#else
+    /* GCLK Setup */
+    GCLK->PCHCTRL[ADC_GCLK_ID].reg = GCLK_PCHCTRL_CHEN | GCLK_PCHCTRL_GEN(ADC_GCLK_SRC);
     /* Configure CTRLB & CTRLC Register */
-    ADC->CTRLB.reg = ADC_PRESCALER;
-    ADC->CTRLC.reg |= res;
+    ADC_DEV->CTRLB.reg = ADC_PRESCALER;
+    ADC_DEV->CTRLC.reg |= res;
+#endif
```
Can you move this block to a `_setup_clock()` helper function

janosbrodbeck replied:
Done.

benpicco commented on cpu/sam0_common/periph/adc.c:
```diff
@@ -120,33 +134,73 @@ static int _adc_configure(adc_res_t res)
+    ADC_DEV->INTENCLR.reg = ADC_INTENCLR_WINMON | ADC_INTENCLR_OVERRUN |
                               ADC_INTENCLR_RESRDY;
     /* Set default calibration from NVM */
-    ADC->CALIB.reg =
+#ifdef ADC0_FUSES_BIASCOMP_ADDR
+    if (ADC_DEV == ADC0) {
+        ADC_DEV->CALIB.reg =
+                ADC0_FUSES_BIASCOMP((*(uint32_t*)ADC0_FUSES_BIASCOMP_ADDR)) >>
+                ADC_CALIB_BIASCOMP_Pos |
+                ADC0_FUSES_BIASREFBUF((*(uint32_t*)ADC0_FUSES_BIASREFBUF_ADDR) >>
+                ADC0_FUSES_BIASREFBUF_Pos);
+    }
+    else {
+        ADC_DEV->CALIB.reg =
+            ADC1_FUSES_BIASCOMP((*(uint32_t*)ADC1_FUSES_BIASCOMP_ADDR)) >>
+            ADC_CALIB_BIASCOMP_Pos |
+            ADC1_FUSES_BIASREFBUF((*(uint32_t*)ADC1_FUSES_BIASREFBUF_ADDR) >>
+            ADC1_FUSES_BIASREFBUF_Pos);
+    }
+#else
+    ADC_DEV->CALIB.reg =
             ADC_FUSES_BIASCOMP((*(uint32_t*)ADC_FUSES_BIASCOMP_ADDR)) >>
             ADC_CALIB_BIASCOMP_Pos |
             ADC_FUSES_BIASREFBUF((*(uint32_t*)ADC_FUSES_BIASREFBUF_ADDR) >>
             ADC_FUSES_BIASREFBUF_Pos);
+#endif
     while (_adc_syncing()) {}
```
Please move this to a `_setup_calibration` helper function.

benpicco commented on cpu/samd5x/include/periph_cpu.h:
```diff
@@ -83,6 +83,19 @@ enum {
+    ADC_RES_16BIT = 0xfd                        /**< not supported */
+} adc_res_t;
+/** @} */
+#endif /* DOXYGEN */
```
You could add a

```C
#define ADC_GCLK_SRC_DEFAULT  SAM0_GCLK_TIMER
```

I don't think the ADC will run off a 120 MHz clock.

benpicco replied:
Ah I see you have a prescaler of 256 in the board config.
So the ADC is running at 468 kHz which albeit a bit slow, should be a valid configuration.

janosbrodbeck replied:
I'll lower the prescaler to 128. Is a bit faster and still runs.

benpicco commented on cpu/sam0_common/periph/adc.c:
```diff
@@ -28,6 +28,20 @@
+#endif
+
+#ifndef ADC_GCLK_SRC
+    #define ADC_GCLK_SRC SAM0_GCLK_MAIN
```
```suggestion
    #define ADC_GCLK_SRC ADC_GCLK_SRC_DEFAULT
```

janosbrodbeck commented:
Murdock showed that it still fails for some boards. I'll fix it.

maribu commented:
It's about time I revisit and complete my `adc_ng` PR. With that API (which has an explicit parameter to select an ADC device), it would be trivial to support both ADCs simultaneously.

janosbrodbeck commented:
> It's about time I revisit and complete my `adc_ng` PR. With that API (which has an explicit parameter to select an ADC device), it would be trivial to support both ADCs simultaneously.

Yes, I had already seen the upcoming approach with adc_ng. But it looked like it would still need some time to finish. Until then we have at least support to use at least one device with these CPUs :)

benpicco commented on cpu/sam0_common/periph/adc.c:
```diff
@@ -120,33 +134,73 @@ static int _adc_configure(adc_res_t res)
-    /* Set Voltage Reference */
-    ADC->REFCTRL.reg = ADC_REF_DEFAULT;
+    if (ADC_DEV == ADC0) {
+        GCLK->PCHCTRL[ADC0_GCLK_ID].reg = GCLK_PCHCTRL_CHEN |
+            GCLK_PCHCTRL_GEN(ADC_GCLK_SRC);
```
```suggestion
        GCLK->PCHCTRL[ADC0_GCLK_ID].reg = GCLK_PCHCTRL_CHEN
                                        | GCLK_PCHCTRL_GEN(ADC_GCLK_SRC);
```

looks a bit nicer imho

janosbrodbeck replied:
The bitwise OR was already in the line before the line break in some other parts of the driver. 

To get the bit operations directly one below the other is unfortunately only possible in 2-3 places.I tried to keep it consistent with the rest of the code.

But I've moved the bitwise or into the next line where it made sense for me now. If it still doesn't quite fit, just let me know.

janosbrodbeck commented:
Murdock passed now. But I could only test it with an saml21-xpro and same54-xpro for now. The problem with the samr21-xpro and saml11-xpro, which is why Murdock failed before, was however a clear #ifdef/#else mistake by me. 

Moved the calibration and clock setup part into helper functions.  

Re-review would be nice :)

benpicco commented on cpu/sam0_common/periph/adc.c:
```diff
@@ -77,6 +93,75 @@ static void _adc_poweroff(void)
 }
 
+static void _setup_clock(adc_res_t res)
+{
```
```suggestion
{
    sam0_gclk_enable(ADC_GCLK_ID);
```

benpicco commented on cpu/sam0_common/periph/adc.c:
```diff
@@ -77,6 +93,75 @@ static void _adc_poweroff(void)
+    GCLK->CLKCTRL.reg = (uint32_t)(GCLK_CLKCTRL_CLKEN
+            | GCLK_CLKCTRL_GEN(ADC_GCLK_SRC) | (GCLK_CLKCTRL_ID(ADC_GCLK_ID)));
+    /* Configure CTRLB Register HERE IS THE RESOLUTION SET! */
+    ADC_DEV->CTRLB.reg = ADC_PRESCALER | res;
```
Not part of the clock setup.

Also, `ADC_DEV->CTRLB.reg = ADC_PRESCALER;` can be set unconditionally, then you'd have

```C
#ifdef ADC_CTRLC_RESSEL
    ADC_DEV->CTRLC.reg |= res;
#else 
    ADC_DEV->CTRLB.reg |= res;
#endif
```

(Also setting `res` directly instead of using `ADC_CTRLC_RESSEL(res)`?)

benpicco replied:
Ah `ADC_RES_10BIT` is already defined to `ADC_CTRLB_RESSEL_10BIT` etc so we don't need to use the helper macro here.

benpicco commented on cpu/sam0_common/periph/adc.c:
```diff
@@ -87,29 +172,20 @@ static int _adc_configure(adc_res_t res)
     }
 #ifdef CPU_SAMD21
     /* Power On */
     PM->APBCMASK.reg |= PM_APBCMASK_ADC;
```
Push that into `_setup_clock()` and you have a section of common code here.

benpicco commented on cpu/sam0_common/periph/adc.c:
```diff
@@ -87,29 +172,20 @@ static int _adc_configure(adc_res_t res)
+    ADC_DEV->REFCTRL.reg = ADC_REF_DEFAULT;
     /* Disable all interrupts */
-    ADC->INTENCLR.reg = (ADC_INTENCLR_SYNCRDY) | (ADC_INTENCLR_WINMON) |
+    ADC_DEV->INTENCLR.reg = (ADC_INTENCLR_SYNCRDY) | (ADC_INTENCLR_WINMON) |
                               (ADC_INTENCLR_OVERRUN) | (ADC_INTENCLR_RESRDY);
```
Since we want to disable all interrupts here anyway, we might as well

```C
    ADC_DEV->INTENCLR.reg = 0xFF;
```

which again works on all members of the family.

benpicco commented on cpu/sam0_common/periph/adc.c:
```diff
@@ -120,33 +196,32 @@ static int _adc_configure(adc_res_t res)
     /* Power on */
 #ifdef CPU_SAML1X
     MCLK->APBCMASK.reg |= MCLK_APBCMASK_ADC;
+#else
+#ifdef MCLK_APBDMASK_ADC0
+    if (ADC_DEV == ADC0) {
+        MCLK->APBDMASK.reg |= MCLK_APBDMASK_ADC0;
+    } else {
+        MCLK->APBDMASK.reg |= MCLK_APBDMASK_ADC1;
+    }
 #else
     MCLK->APBDMASK.reg |= MCLK_APBDMASK_ADC;
```
move that into `_setup_clock()`

janosbrodbeck commented:
I've moved the mentioned parts. Added sam0_gclk_enable at the beginning of _setup_clock(). 

Also made some changes to the formatting of the #ifdef/#else parts. Hope it is okay.

benpicco commented on cpu/sam0_common/periph/adc.c:
```diff
@@ -87,66 +189,38 @@ static int _adc_configure(adc_res_t res)
-                              (ADC_INTENCLR_OVERRUN) | (ADC_INTENCLR_RESRDY);
+    ADC_DEV->INTENCLR.reg = 0xFF;
+#ifdef CPU_SAMD21
     while (_adc_syncing()) {}
```
You can move the `_adc_syncing()` out of the `ifdef` (or remove it entirely unless there is a reason why we have to sync before enabling the band gap. There is a sync at the end after enable, that should be enough.)

janosbrodbeck replied:
I've moved it out of the ifdef. At least in the `ifdef` it was unnecessary to have it in both paths at the beginning.

benpicco commented on cpu/sam0_common/periph/adc.c:
```diff
@@ -177,17 +251,17 @@ int32_t adc_sample(adc_t line, adc_res_t res)
     }
 #ifdef CPU_SAMD21
-    ADC->INPUTCTRL.reg = ADC_GAIN_FACTOR_DEFAULT |
+    ADC_DEV->INPUTCTRL.reg = ADC_GAIN_FACTOR_DEFAULT |
```
No gain on non-samd21 MCUs?

We could just 

```C
#ifndef ADC_GAIN_FACTOR_DEFAULT
#define ADC_GAIN_FACTOR_DEFAULT (0)
#endif
```

and remove the ifdef here.
Since 0 maps to `ADC_INPUTCTRL_GAIN_1X` that should be a sensible default.

janosbrodbeck replied:
Should be sensible, yes. Done!

benpicco commented on cpu/sam0_common/periph/adc.c:
```diff
@@ -48,11 +68,11 @@ static inline void _done(void)
+    if (ADC_DEV->SYNCBUSY.reg) {
         return true;
     }
 #endif
```
All the other peripherals have something like

```C
static inline void _wait_syncbusy(void)
{
#ifdef ADC_STATUS_SYNCBUSY
    while (ADC_DEV->STATUS.reg & ADC_STATUS_SYNCBUSY) {}
#else
    while (ADC_DEV->SYNCBUSY.reg) {}
#endif
}
```

might as well use that here while you're at it.

janosbrodbeck replied:
implemented it. In this way the explicit CPU differentiation is also eliminated, thats nice.

benpicco commented on cpu/sam0_common/periph/adc.c:
```diff
@@ -176,18 +253,14 @@ int32_t adc_sample(adc_t line, adc_res_t res)
-    ADC->INPUTCTRL.reg = adc_channels[line].muxpos | ADC_NEG_INPUT;
-#endif
     while (_adc_syncing()) {}
     /* Start the conversion */
```
```suggestion

    /* Start the conversion */
```

separating the blocks with a newline helps readability

benpicco commented on cpu/sam0_common/periph/adc.c:
```diff
@@ -87,66 +193,37 @@ static int _adc_configure(adc_res_t res)
         SUPC->VREF.reg |= SUPC_VREF_VREFOE;
     }
 #endif
     /*  Enable ADC Module */
```
```suggestion

    /*  Enable ADC Module */
```

benpicco commented on cpu/sam0_common/periph/adc.c:
```diff
@@ -63,20 +83,106 @@ static void _adc_poweroff(void)
+    #ifdef ADC0_FUSES_BIASCOMP_ADDR
+        if (ADC_DEV == ADC0) {
+            ADC_DEV->CALIB.reg =
+                ADC0_FUSES_BIASCOMP((*(uint32_t*)ADC0_FUSES_BIASCOMP_ADDR)) >>
+                ADC_CALIB_BIASCOMP_Pos |
```
Shouldn't that be
```suggestion
                ADC0_FUSES_BIASCOMP((*(uint32_t*)ADC0_FUSES_BIASCOMP_ADDR)) << ADC_CALIB_BIASCOMP_Pos
```

(Also keep the shift on a single line to avoid confusion. 80 Characters are just a *recommendation*.

benpicco replied:
Ah this is to 're-normalize' the fuse value 

Looking closer this evaluates to 

```C
(ADC_FUSES_BIASCOMP_Msk & ((value) << ADC_FUSES_BIASCOMP_Pos)) >>
                ADC_CALIB_BIASCOMP_Pos
```

So while `ADC0_FUSES_BIASCOMP_ADDR & 0x7` would be shorter, using the macros keeps it future proof. (and the compiler will optimize it anyway)

benpicco commented on cpu/sam0_common/periph/adc.c:
```diff
@@ -87,67 +187,43 @@ static int _adc_configure(adc_res_t res)
+    ADC_DEV->CTRLC.reg &= ~ADC_CTRLC_RESSEL_Msk;
+    ADC_DEV->CTRLC.reg |= res;
+#else
+    ADC_DEV->CTRLB.reg |= res;
```
You want to do the same here

janosbrodbeck replied:
No, we set CTRLB explicitly to `ADC_PRESCALER` in `_setup_clock`. It's not necessary here.

janosbrodbeck replied:
True. Thought it is not necessary because we set CTRLB to `ADC_PRESCALER` explicitly in `_setup_clock`. But that's just for those CPUs with CTRLC for the resolution. Missed that one.

benpicco commented:
On `saml21-xpro` only the ADC line on PA02 works, but that is also true on `master`.
On `saml10-xpro` the measurements are reversed (0V is 1024, 3.3V is 0) but again, also on `master`.

janosbrodbeck commented:
> On `saml21-xpro` only the ADC line on PA02 works, but that is also true on `master`.
> On `saml10-xpro` the measurements are reversed (0V is 1024, 3.3V is 0) but again, also on `master`.

Can't confirm the saml10-xpro reversed behaviour. In my test: 0V is 0 and 3.3 V is 1024
